merkletree/merkletree.go

```diff
@@ -0,0 +1,288 @@
+// merkletree provides tools for calculating the Merkle root of a dataset, for
+// creating a proof that a piece of data is in a Merkle tree of a given root,
+// and for verifying proofs that a piece of data is in a Merkle tree of a given
+// root.
+package merkletree
+
+import (
+	"bytes"
+	"hash"
+)
+
+// A Tree takes data as leaves and returns the merkle root. Each call to 'Push'
+// adds one leaf to the merkle tree. Calling 'Root' returns the Merkle root.
+// The Tree also constructs proof that a single leaf is a part of the tree. The
+// leaf can be chosen with 'SetIndex'.
+type Tree struct {
+	head *subTree
+	hash hash.Hash
+
+	// Variables to help build proofs that the data at 'proveIndex' is in the
+	// merkle tree.
+	currentIndex int
+	proveIndex   int
+	proveSet     [][]byte
+}
+
+// A subTree is a subTree in the Tree. 'height' refers to how tall the subTree
+// is. The children of the tree are not accessible, as they have already been
+// hashed into 'sum'. 'next' is the next subTree, and is guaranteed to have
+// a higher height unless it is nil.
+
+// A subTree contains the merkle root of a complete (2^n leaves) subTree of
+// the Tree. 'sum' is the Merkle root of the subTree. If 'next' is not nil, it
+// will be a tree with a higher height.
+type subTree struct {
+	next   *subTree
+	height int
+	sum    []byte
+}
+
+// sum returns the hash of the input data.
+func sum(h hash.Hash, data []byte) []byte {
+	if data == nil {
+		return nil
+	}
+
+	h.Write(data)
+	result := h.Sum(nil)
+	h.Reset()
+	return result
+}
+
+// join takes two byte slices, appends them, hashes them, and then returns the
+// result.
+func join(h hash.Hash, a, b []byte) []byte {
+	return sum(h, append(a, b...))
+}
+
+// New initializes a Tree with a hash object, which will be used when hashing
+// the input.
+func New(h hash.Hash) *Tree {
+	return &Tree{
+		hash: h,
+	}
+}
+
+// Reset returns the tree to its inital, empty state.
+func (t *Tree) Reset() {
+	t.head = nil
+	t.currentIndex = 0
+	t.proveIndex = 0
+	t.proveSet = nil
+}
+
+// SetIndex resets the tree, and then sets the index for which a proof that the
+// element is in the Tree will be built.
+func (t *Tree) SetIndex(i int) {
+	t.Reset()
+	t.proveIndex = i
+}
```

Comment:
The problem is that if you SetIndex to an index that's already passed, you can't do anything.

We could have it return an error if the tree hasn't been Reset(), that might be better.